OpenRA.Mods.Common/Traits/Attack/AttackBase.cs

```diff
@@ -80,6 +80,10 @@ protected virtual bool CanAttack(Actor self, Target target)
 			if (!HasAnyValidWeapons(target))
 				return false;
 
+			var layer = self.Location.Layer;
+			if (layer == CustomMovementLayerType.Tunnel || layer == CustomMovementLayerType.Subterranean)
```

Comment:
Actually, scratch that.  I tested the older version and both that and my earlier AttackFrontal changes do the same thing.  :+1: for the current approach if you can `TraitOrDefault<Mobile>` and check `CanInteractWithGroundLayer`.